### ISSUE TITLE
clang-tidy is broken

### ISSUE THREAD

**Issue by derekargueta** [envoyproxy/envoy#9512]
There are currently 2 issues with clang-tidy:
1. When clang-tidy has an error, the CI job still reports back "pass"
2. clang-tidy *always* has errors on every PR such as the following:
`error: invalid argument '-std=c++14' not allowed with 'C' [clang-diagnostic-error]`
`error: cannot use 'throw' with exceptions disabled [clang-diagnostic-error]`

I'm still troubleshooting issue 1. [This PR](https://github.com/envoyproxy/envoy/pull/9140) added a fix (`-o pipefail`) to unmask an error in `git clone`, but for some reason errors in clang-tidy itself are still being swallowed.

For issue 2 I'm suspecting that it's caused by [this change](https://github.com/envoyproxy/envoy/pull/8628/files#diff-bd94107c801983d39bc0bd34030f0b12L62) in `gen_compilation_database.py` which removes the hard-coded `clang++` compiler invocation in favor of whatever compiler the Bazel magic returns. From observing my `compile_commands.json`, it appears to be the value of `$CC`, not `$CXX` (i.e. I get `gcc` instead of `g++`) which explains why clang-tidy doesn't like exceptions or `-std=c++14`. Interestingly, `-x c++` is correctly set, although I believe to correctly compile C++ with a C compiler you also need to link the stdlib, such as `-lc++`.

cc @lizan

**Comment by derekargueta** [comment]
For issue 2, poking around at clang-tidy outputs, I'm growing suspicious that the Bazel upgrade to 1.1.0 may be the culprit - there's numerous runs after the above-mentioned `gen_compilation_database.py` change where clang-tidy runs as expected. But around the Bazel upgrade is where I'm seeing `error: invalid argument '-std=c++14' not allowed with 'C' [clang-diagnostic-error]` and it's the only build-related change in that neighborhood of commits (circa Oct. 28th)

**Comment by derekargueta** [comment]
And for issue 1, looks like it was the upgrade to LLVM 9.

clang-tidy-diff.py in LLVM 8 used `sys.exit` to exit with the same error code that clang-tidy exits with ([source](https://github.com/llvm/llvm-project/blob/19a71f6bdf2dddb10764939e7f0ec2b98dba76c9/clang-tools-extra/clang-tidy/tool/clang-tidy-diff.py#L143)) whereas in LLVM 9 that's no longer the case - clang-tidy-diff.py always exits with 0 unless a Python exception occurred, which wouldn't happen for clang-tidy errors. ([source](https://github.com/llvm/llvm-project/blob/master/clang-tools-extra/clang-tidy/tool/clang-tidy-diff.py#L263)). It appears to have been removed as a part of [this commit](https://github.com/llvm/llvm-project/commit/d3939d2013e9689f2e85494ee1dfd97dd6a1b76b#diff-fae0f744efff7cd862fd2cf9fcebfa26L142). :sadpanda:

**Comment by mattklein123** [comment]
Thanks for raising this @derekargueta. I think @lizan brought this up recently but to your point it hasn't been fixed yet. We should fix this ASAP and cleanup any issues that have snuck in on master.

**Comment by derekargueta** [comment]
So issue 2 will need to be solved first, because if we fix 1 first, then all CI will just fail clang-tidy, but for issue 1 LLVM is open to a patch to restore the exit code ([communications](https://reviews.llvm.org/D57662#1797607)) but that won't be accessible until the next LLVM release without doing something odd/hacky like pulling that specific script from github, so in the mean time we can run `git diff` ourselves to get a list of diff'd files and run `run-clang-tidy.py` against that set, which has the [correct exit functionality](https://github.com/llvm/llvm-project/blob/master/clang-tools-extra/clang-tidy/tool/run-clang-tidy.py#L323) and is also parallelized. A minor inconvenience but isn't too bad.

**Comment by mattklein123** [comment]
@derekargueta I know you are working on this in your spare time, but would you mind a short summary of what you know about what is broken and what needs to be fixed? I'm wondering if we can get someone to help push this over the finish line if there is something useful that can be done. Thank you!

**Comment by derekargueta** [comment]
Haven't gotten full RCA yet, extra eyes/hands would be much appreciated.

TL;DR, today if you generate compile_commands.json, `clang-tidy` doesn't seem to like it. There have been multiple changes that could affect this since this broke around mid-October. The most significant events (AFAIK):
* Oct. 15, 2019 - [build: use clang-9](https://github.com/envoyproxy/envoy/pull/8613) <-- this broke clang-tidy reporting so that it returns "ok" even if things broke
* Oct. 16, 2019 - [build: make compilation database generation faster](https://github.com/envoyproxy/envoy/pull/8628) <-- Not quite sure if this broke CI - on my system I have to keep the hard-coded `clang++`
* Oct. 23, 2019 - [Update dependencies - Go, Bazel toos, xxHash, nanopb, rules_go, curl, protobuf, Jinja, yaml-cpp, re2](https://github.com/envoyproxy/envoy/pull/8728) <-- this also bumped bazel-compilation-database, but I don't think this change was harmful
* Oct. 28, 2019 - [build: use Bazel 1.1.0](https://github.com/envoyproxy/envoy/pull/8739) <-- this changed the output flags in `compile_commands.json`

I'm most suspicious of the Bazel 1.1.0 upgrade affecting the output of bazel-compilation-database. One of the last sets of flags that it writes out are the builtin include directories ([code here](https://github.com/grailbio/bazel-compilation-database/blob/master/aspects.bzl#L275)). I noticed after Bazel 1.1.0 these are now missing from compile_commands.json. Manually adding in my clang installation's include (`-isystem /usr/local/opt/llvm/lib/clang/9.0.1/include`) to `compile_commands.json` makes clang-tidy happy.

This led me digging into what changed from Bazel 0.29.1 to Bazel 1.1.0 (since that's the jump we made). The most suspicious thing I could find is that [this change](https://github.com/bazelbuild/bazel/commit/74ee2b2e14c12caed7bda1a9c291e19b8f73f5fd#diff-f321a84e135b6943c822c7a94df47b82R249) appears off in that commit - seems it should be `include_directories`, not `framework_include_directories` which is also used a few lines below for `frameworkIncludeDirs`. If `cc_toolchain.built_in_include_directories` is tied to `includeDirs` then that may explain why Bazel 1.1.0 resulted in fewer includes in compile_commands.json. So I left off spelunking in Bazel reading the code and trying to figure out how to test and validate this hypothesis.

--------
appendix
the order of items in the compiler command written by bazel-compilation-database are
1. compiler command itself (e.g. `clang++`)
2. compiler options from `cc_common.get_memory_inefficient_command_line` (e.g. `-fstack-protector`)
3. `compilation_context.defines` (e.g. `-DSPDLOG_FMT_EXTERNAL`)
4. `compilation_context.system_includes` (e.g. `-isystem external/com_github_fmtlib_fmt/include`)
5. `compilation_context.includes` (e.g. `-I bazel-out/darwin-fastbuild/bin/include/envoy/http/_virtual_includes/async_client_interface`)
6. `compilation_context.quote_includes` (e.g. `-iquote external/com_google_absl`)
7. `compiler_info.compile_flags`
8. `cc_toolchain.built_in_include_directories` <-- this went missing after Bazel 1.1.0
9. `force_language_mode_option` i.e. `-x c++`
10. `-c src.path`
(see [aspects.bzl:270](https://github.com/grailbio/bazel-compilation-database/blob/master/aspects.bzl#L270) which brings the above together)

**Comment by mattklein123** [comment]
Thanks a ton @derekargueta for the detailed writeup. cc @marcomagdy ^ if you have any cycles to help sort this out that would be highly appreciated.

**Comment by marcomagdy** [comment]
@derekargueta I wonder if this is related to how the compilation database gets screwed up after running `proto_format.sh`. At any rate I'll take a look.

**Comment by lizan** [comment]
Yes I think that's the cause. It is also affecting my local dev experience. I will try fix that soon.
